### DISCUSSION TITLE
Reading in M values using geopandas

### DISCUSSION THREAD

**Discussion by AnneEstoppey** [geopandas/geopandas#2714] [Q&A]
This is about geodatabase, feature class, geopandas and geojson.

Our feature class is lines. The geometry parameters are X, Y, Z and also M.
M values can be used to store distance between vertices, or some other attribute.
M values are part of the geometry.
We are interested in preserving M values for our lines.

When exporting (or writing) a line feature class into a geojson file, using geopandas, we noticed that M values were missing.

After further investigation, we noticed that in fact the M values are not _read_ in by geopandas.
Displaying the geometry field from the geodataframe showed that M values were dropped.
Geometry parameters shown in the geodataframe are: X, Y and Z.

We could not see in the geopandas documentation that there is an option to add (or drop) M values.

Are we missing something or is the option to read in M values with geopandas not possible (yet)?

If this option does not exist, will it be a future enhancement?

Cheers
Anne

**Comment by jorisvandenbossche** [comment]
This is indeed not supported at the moment. There are several aspects about this, though:

- Read/write support for GIS file formats with M values: we currently rely on the `fiona` library for reading files with GDAL, and fiona doesn't (yet) have support for M values (https://github.com/Toblerity/Fiona/issues/774, https://github.com/Toblerity/Fiona/pull/778)
- Even if we can read M values, we also need to store it, for which we use `GEOS` through the Shapely or PyGEOS packages. I also think here the support is quite limited (see eg https://gis.stackexchange.com/questions/268694/how-to-store-the-shapefile-m-attribute-in-geopandas-shapely, GEOS does not have support for M values:https://trac.osgeo.org/geos/ticket/1089)
- And even if GEOS would gain the support to store M values, preserving/handling them correctly in all spatial operations is yet another thing.

Now, a minimal feature that might be worth looking into (and which does not depend on GEOS supporting M values), is to be able to read the M values into a separate attribute field (i.e. a column in the GeoDataFrame). That is still something that requires support for this in fiona (or pyogrio), but might be more realistic to achieve.

**Comment by AnneEstoppey** [comment]
Hello Joris and thanks a lot for your answer and the links you provided.
This is valuable information as it confirms what we were thinking: M values integration is, so far, missing in Fiona. And therefore Geopandas.
We will definitely establish a work around, I guess we could use WKT format to extract X, Y, Z and M values, and then find "some way" to reinject M values into geometry in our geojson 'out' file. That might introduce quite a bit of extra run time for our ETL routines though, so we'll see about that.

Cheers and hope to meet you soon again :-)

**Comment by dunha** [comment]
It's a shame this has been left for so long. GEOS claim to support simple features and be a port of JTS, but simple features support M values and JTS also has for 5 or 6 years at v 1.16. I note that it has been moved to a later milestone
The idea of supporting Ms as a separate field could work, considering ESRI appears to support them with an array of doubles in shapefiles and FileGDBs

**Comment by tariqshihadah** [comment]
I've been working on developing a package that builds on geopandas and others that offers some support for linear data with M values, though it cannot read these directly from, for example, a route layer in a GDB. It primarily reads this through begin/end locators included in separate fields in a geodataframe. The package does support a wide variety of operations for linearly referenced data, such as dissolves, merges, overlays, sliding window analysis, etc.

The package is called **linref** and is currently being used by a small group but is working well! Plenty of room for growth and additional features.

More info about it at: [https://linref.readthedocs.io/en/latest/](https://linref.readthedocs.io/en/latest/)
